### PR TITLE
[Kafka]Kafka sink numRecordsOutByIncremental metrics

### PR DESCRIPTION
**Pull request by ChengJie1053**
numRecordsOutByIncremental metrics collects incremental statistics

numRecordsOutByIncremental has three metrics, respectively is numRecordsOutByIncrementalInsert, numRecordsOutByIncrementalUpdate, numRecordsOutByIncrementalDelete

image

2024.7.20
image

### CODE REVIEW

File: flink-cdc-connect/flink-cdc-pipeline-connectors/flink-cdc-pipeline-connector-kafka/src/main/java/org/apache/flink/cdc/connectors/kafka/sink/KafkaSinkWriteMetrics.java [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+    private final Map<String, Counter> numRecordsOutMap = new ConcurrentHashMap();
+
+    public KafkaSinkWriteMetrics(MetricGroup metricGroup) {
+        this.metricGroup = metricGroup;
+    }
+
+    public void numRecordsOutInsertIncrease(TableId tableId) {
+        Counter counter =
+                numRecordsOutMap.computeIfAbsent(
+                        tableId.identifier() + OperationType.INSERT,
```
Comment (yuxiqian) [inline flink-cdc-connect/flink-cdc-pipeline-connectors/flink-cdc-pipeline-connector-kafka/src/main/java/org/apache/flink/cdc/connectors/kafka/sink/KafkaSinkWriteMetrics.java:51]: As previously pointed out by @lvyanquan, identifier string generating & hashing could be expensive considering it must be done for every DataChangeEvent. What about splitting `INSERT`, `UPDATE`, and `DELETE` counters into three `Map<TableId, Counter>` hash map, so the `TableId` key hash code could be safely cached after we merge #3454?

Comment (ChengJie1053) [reply flink-cdc-connect/flink-cdc-pipeline-connectors/flink-cdc-pipeline-connector-kafka/src/main/java/org/apache/flink/cdc/connectors/kafka/sink/KafkaSinkWriteMetrics.java:51]: Ok, thank you for helping me review the code, I will modify this part

**Comment by lvyanquan** [comment]
I have two points to be discussed.
1. Does 'numRecordsOutByIncremental' refer to the incremental phase here? We don't have a way to determine whether we have reached this phase in the sink now.
2. If we want to add these metrics, I suggest adding it uniformly though [DataSinkWriterOperator](https://github.com/apache/flink-cdc/blob/94a0415475e7040cc6a61e9bd8aca9c3ccea80ad/flink-cdc-runtime/src/main/java/org/apache/flink/cdc/runtime/operators/sink/DataSinkWriterOperator.java#L63) for all pipeline sinks.

**Comment by ChengJie1053** [comment]
> I have two points to be discussed.
> 
> 1. Does 'numRecordsOutByIncremental' refer to the incremental phase here? We don't have a way to determine whether we have reached this phase in the sink now.
> 2. If we want to add these metrics, I suggest adding it uniformly though [DataSinkWriterOperator](https://github.com/apache/flink-cdc/blob/94a0415475e7040cc6a61e9bd8aca9c3ccea80ad/flink-cdc-runtime/src/main/java/org/apache/flink/cdc/runtime/operators/sink/DataSinkWriterOperator.java#L63) for all pipeline sinks.

Ok, thanks for reviewing the code for me
1.`numRecordsOutByIncremental` the name may be ambiguous, I will modify the name
2.I adding it uniformly though DataSinkWriterOperator for all pipeline sinks.

**Pull request closed without merging**